### ISSUE TITLE
NeuralNetworkRegressor models and BatchNorm layers

### ISSUE THREAD

**Issue by Leonardbcm** [FluxML/MLJFlux.jl#142]
Hi everyone,

I tired using the `NeuralNetworkRegressor` model. I designed the following `builder`: 

```julia
using  MLJ, MLJFlux, Flux, DataFrames
mutable struct TESTBuilder <: MLJFlux.Builder end
MLJFlux.build(builder::TESTBuilder, n_in, n_out) = Chain(BatchNorm(n_in), Dense(n_in, n_out))
```

which builds a very simple model:

```julia
n = 100
d = 5
builder = TESTBuilder()
MLJFlux.build(builder, n, d)
```
>>> Chain(BatchNorm(100), Dense(100, 5))

Let's try to train this model: 

```julia
# We generate dummy data in the appropriate format
X = DataFrame(rand(Float64, n, d), :auto);
y = rand(Float64, n);

Model = @load NeuralNetworkRegressor pkg=MLJFlux verbosity=0
model = Model(builder = TESTBuilder(), batch_size=25)
mach = machine(model, X, y)
fit!(mach)
```
Everything seems to have worked fine.

>>> [ Info: Training Machine{NeuralNetworkRegressor{TESTBuilder,…},…} @137.                                                                                                                                            Optimising neural net:100%[=========================] Time: 0:00:36                                                                                                                                                Machine{NeuralNetworkRegressor{TESTBuilder,…},…} @137 trained 1 time; caches data                                                                                                                                    args:                                                                                                                                                                                                                1:  Source @542 ⏎ `Table{AbstractVector{Continuous}}`                                                                                                                                                              2:  Source @578 ⏎ `AbstractVector{Continuous}`  

However, it's while trying to use the `predict` method that things mess up:
```julia 
predict(mach, X)
```
>>> ERROR: arraysize: dimension out of range                                                                                                                                                                           Stacktrace:                                                                                                                                                                                                          [1] size                                                                                                                                                                                                             @ .\array.jl:132 [inlined]                                                                                                                                                                                       [2] (::BatchNorm{typeof(identity), Vector{Float32}, Float32, Vector{Float32}})(x::Vector{Float64})                                                                                                                   @ Flux ~\.julia\packages\Flux\6BByF\src\layers\normalise.jl:269                                                                                                                                                  [3] applychain(fs::Tuple{BatchNorm{typeof(identity), Vector{Float32}, Float32, Vector{Float32}}, Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}}, x::Vector{Float64})                                     @ Flux ~\.julia\packages\Flux\6BByF\src\layers\basic.jl:36                                                                                                                                                       [4] Chain                                                                                                                                                                                                            @ ~\.julia\packages\Flux\6BByF\src\layers\basic.jl:38 [inlined]                                                                                                                                                  [5] (::MLJFlux.var"#21#23"{Matrix{Float64}, Chain{Tuple{BatchNorm{typeof(identity), Vector{Float32}, Float32, Vector{Float32}}, Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}}}})(i::Int64)              @ MLJFlux .\none:0                                                                                                                                                                                               [6] iterate                                                                                                                                                                                                          @ .\generator.jl:47 [inlined]                                                                                                                                                                                    [7] collect(itr::Base.Generator{UnitRange{Int64}, MLJFlux.var"#21#23"{Matrix{Float64}, Chain{Tuple{BatchNorm{typeof(identity), Vector{Float32}, Float32, Vector{Float32}}, Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}}}}})                                                                                                                                                                                               @ Base .\array.jl:678                                                                                                                                                                                            [8] predict(model::NeuralNetworkRegressor{TESTBuilder, ADAM, typeof(Flux.Losses.mse)}, fitresult::Tuple{Chain{Tuple{BatchNorm{typeof(identity), Vector{Float32}, Float32, Vector{Float32}}, Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}}}, Bool, Vector{String}}, Xnew_::DataFrame)                                                                                                                                       @ MLJFlux ~\.julia\packages\MLJFlux\NTAl9\src\regressor.jl:194                                                                                                                                                   [9] predict(mach::Machine{NeuralNetworkRegressor{TESTBuilder, ADAM, typeof(Flux.Losses.mse)}, true}, Xraw::DataFrame)                                                                                                @ MLJBase ~\.julia\packages\MLJBase\hLtde\src\operations.jl:83                                                                                                                                                  [10] top-level scope                                                                                                                                                                                                  @ REPL[22]:1                                                                                                                                                                                                    [11] top-level scope                                                                                                                                                                                                  @ ~\.julia\packages\CUDA\Px7QU\src\initialization.jl:81   

The error comes from the call to `size` in the assertion of the call of the [BatchNorm](https://github.com/FluxML/Flux.jl/blob/cde0a4590c16d61bd9e2f6a52ccfe48a2d3014cb/src/layers/normalise.jl#L199-L233) layer:
```julia
@assert size(x, ndims(x)-1) == BN.chs
``` 
In fact, the `size` method is called with a second parameter lower than or equal to 0, meaning that the given input `x` is a vector.

Let's have a look at the `predict` method for [NeuralNetworkRegressor](https://github.com/FluxML/MLJFlux.jl/blob/dev/src/regressor.jl) : 
```julia
Xnew_ = MLJModelInterface.matrix(Xnew_)
return [chain(values.(Xnew_[i, :]))[1] for i in 1:size(Xnew_, 1)]
```
I guess the idea is to predict the output 1 by 1 and then concatenate the results with the comprehension. On my data. this gives:
```julia
Xnew_ = MLJModelInterface.matrix(X)
i = 1
v = values.(Xnew_[i, :])
```
>>> 5-element Vector{Float64}:                                                                                                                                                                                          0.9264522698778064                                                                                                                                                                                                 0.6793855588857194                                                                                                                                                                                                 0.2071992861888845                                                                                                                                                                                                 0.12771926540098222                                                                                                                                                                                                0.354197967479825

Which is well a Vector of size `d`, not supported by the `BatchNorm` call function.
Let's try reshaping this vector into the appropriate format : 
```julia
mat_v = zeros(Float64, d, 1)
mat_v[:, 1] = v
mat_v
```
>>> 5×1 Matrix{Float64}:                                                                                                                                                                                                0.9264522698778064                                                                                                                                                                                                 0.6793855588857194                                                                                                                                                                                                 0.2071992861888845                                                                                                                                                                                                 0.12771926540098222                                                                                                                                                                                                0.354197967479825   

Now we have a `d * 1` matrix. 
We directly access the chain object in the fitresults : 
```julia
mach.fitresult[1](mat_v) 
```
>>> 1×1 Matrix{Float64}:                                                                                                                                                                                                -0.7403223392235815 

Which works.

Note that I have not identified a viable way to avoid this problem. Moreover, It might be that `BatchNorm` in `Flux` should allow their methods to work on 1D vectors, in this case sorry for the post.

Please let me know how can we solve this, thanks by advance.

**Comment by ayush-1506** [comment]
@Leonardbcm Thanks for the detailed issue description!
I think the issue is that we should try to avoid `Vector` type inputs to the model. Vectors should be converted to `Nx1 Matrix` everywhere. While most layers support `Vector` type input, others (such as BatchNorm) don't. 

Simple way to reproduce the issue in Flux:

```
ip = rand(5,1)
Dense(5,10)(ip)  #works, outputs 10x1 Matrix

ip2 = ip |> vec # vector
Dense(5,10)(ip2) #works, but output type is Vector
```

But for batchnorm:

```
 BatchNorm(10)(rand(10,1)) # works
 BatchNorm(10)(rand(10,1) |> vec) # Error: arraysize: dimension out of range (because of the size check you mentioned)
```

**Comment by Leonardbcm** [comment]
@ayush-1506 Thanks for the reply. If my understanding is correct, Flux is intended to manipulated `Matrix` types rather than `Vector`. In that case, the error comes from MLFlux's `predict` function which should be corrected.

While I was waiting for an answer, I made this shortcut which redifine the calling behavior of a `BatchNorm` layer in order for it to accept vectors (the `()` function). As I understand it, Flux is not intended to work with vectors and thus this solution should not be preferred. However, here it is : 

```julia
function (BN::BatchNorm)(x)
    x isa Vector && (x = reshape(x, length(x), 1))
    @assert size(x, ndims(x)-1) == BN.chs
    
    N = ndims(x)
    reduce_dims = [1:N-2; N]
    affine_shape = ntuple(i -> i == N-1 ? size(x, N-1) : 1, N)
    return Flux._norm_layer_forward(BN, x; reduce_dims, affine_shape)
end
```

It only consists in adding the first line which reshapes the inputs if it is not correct. Here are some manual tests for checking this solution : 

```julia
#############################
# Test the batch Norm

# Generate data
d = 2
n = 3
bn = BatchNorm(d)
X = zeros(Float64, d, n)
X[1, :] = [1, 2, 3]
X[2, :] = [2, 3, 4]

# While not trained, the batch norm layer acts as a passthrough.
# The small difference is due to the epsilon introduced to stabilize numerical computations.

# Check in matrix mode and in vector mode
bn(X)
[bn(X[:, i]) for i in 1:n]

# Set the moments of the batch norm layer
bn.μ = mean(X; dims=2) |> vec
bn.σ² = std(X; dims=2) |> vec

# Check in matrix mode and in vector mode
bn(X)
[bn(X[:, i]) for i in 1:n]
```

Our modification seems to be working. Let's try to use the predict function.

```julia
using  MLJ, MLJFlux, Flux, DataFrames
mutable struct TESTBuilder <: MLJFlux.Builder end
MLJFlux.build(builder::TESTBuilder, n_in, n_out) = Chain(BatchNorm(n_in), Dense(n_in, n_out))

n = 100
d = 5
builder = TESTBuilder()
MLJFlux.build(builder, n, d)
# We generate dummy data in the appropriate format
X = DataFrame(rand(Float64, n, d), :auto);
y = rand(Float64, n);

Model = @load NeuralNetworkRegressor pkg=MLJFlux verbosity=0
model = Model(builder = TESTBuilder(), batch_size=25)
mach = machine(model, X, y)
fit!(mach)
predict(mach, X)
```

This code is working fine now.

Here is now a second solution, which is modifying the `predict` method for `Regressor` structures in MLJFlux. Don't forget to start a new julia session before trying this solution, otherwise the calling behavior of BatchNorm will still be modified!
```julia
########################## MLJFLUX
using MLJModelInterface
const Regressor = MLJFlux.Regressor
import MLJFlux.predict

function tomat end
tomat(x::Matrix) = x
tomat(x::Vector) = reshape(x, size(x, 1), 1)

function MLJModelInterface.predict(model::Regressor, fitresult, Xnew_)

    chain , target_is_multivariate, target_column_names = fitresult

    Xnew_ = MLJModelInterface.matrix(Xnew_)
    
    if target_is_multivariate
        ypred = [chain(tomat(values.(Xnew_[i, :])))
                 for i in 1:size(Xnew_, 1)]
        return MLJModelInterface.table(reduce(hcat, y for y in ypred)',
                                       names=target_column_names)
    else
        return [chain(tomat(values.(Xnew_[i, :])))[1]
                for i in 1:size(Xnew_, 1)]
    end
end
```

The idea is to define a `tomat` function that does nothing if the data is already a matrix, but reshapes the given vector to a Matrix if the data is a vector. We add a call to this function just before passing the data to the `chain` to make sure the correct format is used.

You can check the results of this solution using the previous code. It would be nice if this kind of solution would be added to MLJFlux, it will improve the interface with Flux.

Thaks for the support

**Comment by ayush-1506** [comment]
> As I understand it, Flux is not intended to work with vectors

@Leonardbcm Right, reshaping is definitely the easiest (and straight-forward) way of getting around this problem from MLJFlux's end. However, I'm wondering if this should ideally be handled at the Flux side. The idea is that we should have uniformity across different layers. Either all layers should support Vector types (and reshape them internally if needed), or all of them should fail and ask the user to convert the Vector to `Nx1 Matrix`.

I'll get in touch with Flux devs to check what they think about this. Either way, would you mind opening a PR for the MLJFlux change?

**Comment by Leonardbcm** [comment]
@ayush-1506 Thanks for your support. Indeed, perhaps the best option would be that Flux uniforms the behavior of their layers, either to authorize all layers to work with matrices, or forbid it.

During the mean time, we can modify MLJFlux with the `tomat` function and the modification of the `predict` function. I will try to run the tests before making the PR.

**Comment by ablaom** [comment]
@Leonardbcm Thanks for raising this issue and your exhaustive analysis.

Your second `tomat` solution makes sense to me. It would be great to get a PR along these lines. Could you also do the same for the NeuralNetworkClassifier  model, where the same issue applies?
